File: concepts/04_collections_and_iteration/06_generators_laziness_and_early_termination/cpp/test_02_closing_and_delegation.cpp
// 提前终止的关闭协议与委托。
// 共同问题：消费方提前停止时是否通知生产方；清理如何穿过委托层；
// 关闭失败如何传播给消费方。
//
// polyglot-family: collections_and_iteration
// polyglot-concept: generators_laziness_and_early_termination
// polyglot-related: languages/cpp/standard_library/09_ranges/
// polyglot-related+: test_085_take_drop_take_while_and_drop_while_prefix_views.cpp
// polyglot-related: languages/cpp/language/test_021_coroutines_promise_awaiter_and_generator.cpp

#include <gtest/gtest.h>

#include <string>
#include <vector>

namespace {

class ClosableRange {
 public:
  explicit ClosableRange(std::vector<std::string>& events) : events_(events) {}

  auto begin() {
    events_.push_back("begin");
    return values_.begin();
  }

  auto end() {
    return values_.end();
  }

  void close() {
    events_.push_back("close");
  }

 private:
  std::vector<std::string>& events_;
  std::vector<int> values_{1, 2};
};

TEST(EarlyTerminationConcept, RangeForBreakDoesNotCallAnAdHocCloseMember) {
  std::vector<std::string> events;
  ClosableRange values{events};

  for (int value : values) {
    EXPECT_EQ(value, 1);
    break;
  }

  EXPECT_EQ(events, (std::vector<std::string>{"begin"}));

  values.close();
  EXPECT_EQ(events, (std::vector<std::string>{"begin", "close"}));

  // range-for 只展开 begin/end、比较、解引用和递增，没有 Python generator.close 或
  // JavaScript IteratorClose。需要提前清理的 range 必须用 RAII 生命周期或显式协议。
}

}  // namespace

File: concepts/04_collections_and_iteration/06_generators_laziness_and_early_termination/julia/test_02_closing_and_delegation.jl
# polyglot-family: collections_and_iteration
# polyglot-concept: generators_laziness_and_early_termination
# polyglot-related: languages/julia/tooling_and_runtime/12_tasks_channels_and_cooperative_concurrency/
# polyglot-related+: test_092_channel_bound_producers_and_iteration.jl
#
# 共同问题：生产者如何声明完成；提前终止是否有通用 close/delegation 协议。
# 对照观察：Julia generator 没有统一 close/yield-from 协议；Channel 用 close 和迭代显式表达生产完成。

using Test

@testset "Channel 的关闭状态终止迭代" begin
    channel = Channel{Int}(2) do output
        put!(output, 1)
        put!(output, 2)
    end
    @test collect(channel) == [1, 2]
    @test !isopen(channel)
    @test_throws InvalidStateException put!(channel, 3)
end
